packages/block-editor/src/components/use-block-commands/index.js

```diff
@@ -0,0 +1,292 @@
+		insertAfterBlock,
+		insertBeforeBlock,
+		flashBlock,
+		setBlockMovingClientId,
+		setNavigationMode,
+		selectBlock,
+	} = useDispatch( blockEditorStore );
+
+	const notifyCopy = useNotifyCopy();
+	const pasteStyles = usePasteStyles();
+
+	const onDuplicate = () => {
+		if ( ! canDuplicate ) {
+			return;
+		}
+		return duplicateBlocks( clientIds, true );
+	};
+	const onRemove = () => {
+		if ( ! canRemove ) {
+			return;
+		}
+		return removeBlocks( clientIds, true );
+	};
+	const onInsertBefore = () => {
+		if ( ! canInsertDefaultBlock ) {
+			return;
+		}
+		const clientId = Array.isArray( clientIds ) ? clientIds[ 0 ] : clientId;
+		insertBeforeBlock( clientId );
+	};
+	const onInsertAfter = () => {
+		if ( ! canInsertDefaultBlock ) {
+			return;
+		}
+		const clientId = Array.isArray( clientIds )
+			? clientIds[ clientIds.length - 1 ]
+			: clientId;
+		insertAfterBlock( clientId );
+	};
+	const onMoveTo = () => {
+		if ( ! canMove ) {
+			return;
+		}
+		setNavigationMode( true );
+		selectBlock( clientIds[ 0 ] );
+		setBlockMovingClientId( clientIds[ 0 ] );
+	};
+	const onGroup = () => {
+		if ( ! blocks.length ) {
+			return;
+		}
+
+		const groupingBlockName = getGroupingBlockName();
+
+		// Activate the `transform` on `core/group` which does the conversion.
+		const newBlocks = switchToBlockType( blocks, groupingBlockName );
+
+		if ( ! newBlocks ) {
+			return;
+		}
+		replaceBlocks( clientIds, newBlocks );
+	};
+	const onUngroup = () => {
+		if ( ! blocks.length ) {
+			return;
+		}
+
+		const innerBlocks = blocks[ 0 ].innerBlocks;
+
+		if ( ! innerBlocks.length ) {
+			return;
+		}
+
+		replaceBlocks( clientIds, innerBlocks );
+	};
+	const onCopy = () => {
+		const selectedBlockClientIds = blocks.map(
+			( { clientId } ) => clientId
+		);
+		if ( blocks.length === 1 ) {
+			flashBlock( selectedBlockClientIds[ 0 ] );
+		}
+		notifyCopy( 'copy', selectedBlockClientIds );
+	};
+	const onPasteStyles = async () => {
+		if ( ! canCopyStyles ) {
+			return;
+		}
+		await pasteStyles( blocks );
+	};
+
+	if ( ! clientIds || clientIds.length < 1 ) {
+		return { isLoading: false, commands: [] };
+	}
+
+	const commands = [
+		onPasteStyles,
+		onCopy,
+		onUngroup,
+		onGroup,
+		onMoveTo,
+		onInsertAfter,
+		onInsertBefore,
+		onRemove,
+		onDuplicate,
+	].map( ( callback ) => {
+		const action = callback.name
+			.replace( 'on', '' )
+			.replace( /([a-z])([A-Z])/g, '$1 $2' );
+		return {
+			name: 'core/block-editor/action-' + callback.name,
+			// translators: %s: type of the command.
+			label: sprintf( __( `%s block` ), action ),
+			icon: copy,
+			callback: ( { close } ) => {
+				callback();
+				close();
+			},
+		};
+	} );
+
+	return { isLoading: false, commands };
+};
+
+export const useBlockCommands = () => {
+	useCommandLoader( {
+		name: 'core/block-editor/blockTransforms',
+		hook: useTransformCommands,
+	} );
+	useCommandLoader( {
+		name: 'core/block-editor/blockActions',
+		hook: useActionsCommands,
+	} );
+};
```

Comment:
Both of these command loaders use a lot of selectors... I do wonder about a potential performance impact on the command palette with long posts. Maybe it's fine though. One follow-up here could be to add a performance metric about the command palette.